### PR TITLE
Portcls fix new

### PR DESCRIPTION
**Pull request by MANIKLEON**
Fix WavePci Port Drivers, (like ac97 sample driver)
Also, fix race conditions

Proposed changes
Split buffers on-page to prevent non-contiguous memory from being passed to the driver
Protect GetMappingWithTag, ReleaseMappingWithTag with spinlock to prevent race conditions
(GetMapping, ReleaseMapping do not need spinlock, they are only called from a service routine)
Remove HasLastMappingFailed(), it never worked properly and can never work

Additional changes:
Move some variables to class level from IRP data, (they did not need to be in IRP data as only one set is active at a time)
Remove m_NumDataAvailable, it was not implemented correctly and was only used for debugging info, if that info is really needed it can be recreated on demand by walking the IRP list. (Not implemented yet)
Removed all remnants of volatile and interlocked operations, they are not needed and were not correct, anyway

### CODE REVIEW

**Comment by binarymaster** [comment]
Duplicate of #3251.

**Comment by MANIKLEON** [comment]
@TheDeadFish hasn't responded to another user's attempt at contacting him for a couple of months. So, I am willing to work on it in his place and want to see if he will ever work on it again. If not, then the continuation of his work can be merged when it is accepted by you all.

**Comment by binarymaster** [comment]
Alright then, but we can't close previous PR unless it's superseded. Consider joining our chat for discussing development.

File: drivers/wdm/audio/backpln/portcls/irpstream.cpp [hunk 1]
```diff
@@ -381,22 +411,19 @@ CIrpQueue::UpdateMapping(
     ASSERT(StreamData);
 
     // add to current offset
-    InterlockedExchangeAdd((PLONG)&m_CurrentOffset, (LONG)BytesWritten);
+    m_CurrentOffset += BytesWritten;
```
Comment (GeoB99) [inline drivers/wdm/audio/backpln/portcls/irpstream.cpp:414]: As you've removed that call, the operation is no longer done in an atomic way to the current offset. Could you please explain the reasoning behind this?

Comment (TheDeadFish) [reply drivers/wdm/audio/backpln/portcls/irpstream.cpp:414]: I am pretty sure GetMapping and UpdateMapping are always called from the same thread.

Comment (TheDeadFish) [reply drivers/wdm/audio/backpln/portcls/irpstream.cpp:414]: If you look at the code for GetMapping, and UpdateMapping, they are definitely not intended to be thread safe. I made hardly any changes to those functions, they are used by wavecyclic, all the signficant changes I made were for the wavepci side. GetMappingWithTag, ReleaseMappingWithTag. CIrpQueue is kind of confusing as it implements two seperate things in the same class.

**Comment by GeoB99** [comment]
@MANIKLEON This is so fishy, are you actually intending to work on this or did _someone else_ (not citing the name) ask you to create a PR so that _someone else_ thinks it'll get merged faster?

**Comment by GeoB99** [comment]
Zero3K, can you please get a job, a life and just move on please? We've endlessly told you to back off already.
Capture

**Comment by MANIKLEON** [comment]
> @MANIKLEON This is so fishy, are you actually intending to work on this or did _someone else_ (not citing the name) ask you to create a PR so that _someone else_ thinks it'll get merged faster?

Hey, @GeoB99  Yeah, I have been fooled by someone on UPWORK. He just wanted to Create a PR request and I didn't know why. I did it for some bucks and he didn't even pay me. 
Sorry to bother please close all requests.

**Comment by learn-more** [comment]
> > @MANIKLEON This is so fishy, are you actually intending to work on this or did _someone else_ (not citing the name) ask you to create a PR so that _someone else_ thinks it'll get merged faster?
> 
> Hey, @GeoB99 Yeah, I have been fooled by someone on UPWORK. He just wanted to Create a PR request and I didn't know why. I did it for some bucks and he didn't even pay me.
> Sorry to bother please close all requests.

Guess it's time to report him then.

**Comment by GeoB99** [comment]
In that case I profusely apologize for accusing you as Zero3K. I'm not sure if you were aware of this but a person called by Zero3K as nickname has been harassing the ReactOS developers and contributors alike into forcing us to do audio related work, and he's been nagging constantly. We'll try to take whatever measures in regard of this.

Sorry again in advance.

**Comment by MANIKLEON** [comment]
> In that case I profusely apologize for accusing you as Zero3K. I'm not sure if you were aware of this but a person called by Zero3K as nickname has been harassing the ReactOS developers and contributors alike into forcing us to do audio related work, and he's been nagging constantly. We'll try to take whatever measures in regard of this.
> 
> Sorry again in advance.

No issues. Same with me constantly messaging to reply to you guys. I Didn't know about this guy (zero3k). Yes, take strict measures.

**Pull request closed without merging**